src/Microsoft.AspNetCore.Mvc.TagHelpers/TagHelperOutputExtensions.cs

```diff
@@ -154,6 +156,149 @@ public static void MergeAttributes(this TagHelperOutput tagHelperOutput, TagBuil
             }
         }
 
+        /// <summary>
+        /// Adds the given <paramref name="classValue"/> to the <paramref name="tagHelperOutput"/>'s
+        /// <see cref="TagHelperOutput.Attributes"/>.
+        /// </summary>
+        /// <param name="tagHelperOutput">The <see cref="TagHelperOutput"/> this method extends.</param>
+        /// <param name="classValue">The class value to add.</param>
+        /// <param name="htmlEncoder">The current HTML encoder.</param>
+        public static void AddClass(
+            this TagHelperOutput tagHelperOutput,
+            string classValue,
+            HtmlEncoder htmlEncoder)
+        {
+            if (tagHelperOutput == null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperOutput));
+            }
+
+            var encodedClassValue = htmlEncoder.Encode(classValue);
+
+            if (SpaceChars.Any(encodedClassValue.Contains))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotContainHtmlSpace, nameof(classValue));
+            }
+
+            if (!tagHelperOutput.Attributes.TryGetAttribute("class", out TagHelperAttribute classAttribute))
+            {
+                tagHelperOutput.Attributes.Add("class", classValue);
+            }
+            else
+            {
+                var currentClassValue = ExtractClassValue(classAttribute, htmlEncoder);
+
+                if (currentClassValue.Equals(encodedClassValue))
+                {
+                    return;
+                }
+
+                if (SpaceChars.Any(currentClassValue.Contains))
```

Comment:
There's a new wrinkle here. It's also because the `HtmlEncoder` is likely to encode the HTML spaces other than `'\u0020'`. Suggest a `private readonly string[]` field containing the encoded strings of each (again, other than `'\u0020'`) and try those too. I say "too" because someone may have slipped an un-encoded HTML space into the attribute's value.

Reply:
So add `private static readonly string[] EncodedSpaceChars = { "&#x9;", "&#xA;", "&#xC;", "&#xD;" };`
And replace `if (SpaceChars.Any(currentClassValue.Contains))` with  `if (SpaceChars.Any(currentClassValue.Contains) || EncodedSpaceChars.Any(currentClassValue.Contains))`

Reply:
Ooops, my mistake. Can't make `EncodedSpaceChars` `static` because that assumes a specific `HtmlEncoder`. It's possible to completely change what gets encoded and how the encoding is done.

So, run the `htmlEncoder` on most `SpaceChars` to initialize an `encodedSpaceChars` variable.

Reply:
Would that mean i have to split on both spaceChars and then encodedSpaceChars?

Reply:
> Would that mean i have to split on both `spaceChars` and then `encodedSpaceChars`?

Unfortunately yes. Can skip it when the `EncodedSpaceChars.Any(...)` check returns `false` but that's about the only shortcut I see.

Reply:
So the sequence of checks would be:
`var containsSpace = SpaceChars.Any(currentClassValue.Contains)`
`var containsEncodedSpace = encodedSpaceChars.Any(currentClassValue.Contains)`

* if containsSpace and doesnt containsEncodedSpace then do the existing workflow
* If it doesnt containsSpace and containsEncodedSpace then split on the encoded space and check if the string is in the array
* if containsSpace and containsEncodedSpace then I need to split on one of them first and then check all the split values for the other one to split on. 

Does normal space split come first over the encoded space split if they are both present?

Reply:
> Does normal space split come first over the encoded space split if they are both present?

Sounds reasonable.